lib/providers/plate_weights.dart

```diff
@@ -0,0 +1,70 @@
+import 'package:flutter/widgets.dart';
+import 'package:wger/helpers/gym_mode.dart';
+
+class PlateWeights extends ChangeNotifier{
+  String unitOfPlate = 'kg';
+  bool flag = false;
+  num totalWeight = 0;
+  num barWeight = 20;
+  num convertTolbs = 2.205;
+  num totalWeightInKg = 0;
+  num barWeightInKg = 20;
+  List<num> selectedWeights = [];
+  List<num> kgWeights = [1.25, 2.5, 5, 10, 15, 20, 25];
+  List<num> lbsWeights = [2.5, 5, 10, 25, 35, 45];
+  List<num> customPlates = [];
+  late Map<num,int> grouped;
+
+  void unitChange() {
+    if(unitOfPlate=='lbs') {
+      totalWeight = totalWeightInKg;
+      unitOfPlate = 'kg';
+      barWeight = barWeightInKg;
+    } else {
+      unitOfPlate = 'lbs';
+      totalWeight = totalWeightInKg*2.205;
+      barWeight = barWeightInKg*2.205;
+    }
+    notifyListeners();
+  }
+
+  void toggleSelection(num x) {
```

Comment:
I would save the selected weights to the device, then we don't need to add this every time. You can simply save the array to SharedPreferences (just dump it as a string and later read from that)

Reply:
I'm currently facing some challenges while implementing shared preferences to manage plate weights within my application. I've integrated the PlateWeightsProvider as intended.

I've explored several approaches:

**Async Function in initState():** I attempted to call an async function containing await loadFromSharedPref() within the initState() of my widget. However, this resulted in widget.getPlates executing before loadFromSharedPref(), leading to an empty list being displayed initially but it is being shown inside the slider selection menu.


**Calling loadFromSharedPref() from main()**: I tried calling loadFromSharedPref() from the main() function to ensure it executed before widget.getPlates. While this successfully loaded and printed the saved values in the main() function, the widget.getPlates still displayed an empty list but it is being shown inside the slider selection menu.

**Flag in PlateWeightsProvider**: I introduced a flag (loadedFromSharedPref) within the PlateWeightsProvider, setting it to true after loadFromSharedPref() completes. In widget.getPlates, I attempted to wait for the flag to become true before displaying the plate denominations. Unfortunately, this approach also failed to display the saved results correctly, and I'm now unable to select/deselect any options or reset them.
I've attached screenshots of the relevant functions for further clarity.

I would greatly appreciate any guidance or suggestions on how to effectively load and display shared preferences within my PlateWeightsProvider and ensure proper functionality within my widgets.

![declared_in_main](https://github.com/user-attachments/assets/b03ce55c-ee9a-4109-be32-b5d2edfc4b33)
![flutter 2](https://github.com/user-attachments/assets/d8940526-a545-433a-8180-3af51e00c111)
![Untitled 3](https://github.com/user-attachments/assets/5e92f711-2b40-4e05-8e13-e4e5b4c55e1b)

https://github.com/user-attachments/assets/1752d933-666a-4a63-aff0-c378d5be73e3

Reply:
mhh, hadn't thought this would cause problems. I'll take a look!

Reply:
@AyushJagaty I got this to work by calling the prefs loader on initstate on the widget (the method itself is also simpler than yours, I think we can get away with just catching some errors and returning an empty list)

in the provider
```dart
  void readPlates() async {
    final prefs = await SharedPreferences.getInstance();
    final platePrefData = prefs.getString(PREFS_PLATES);
    if (platePrefData != null) {
      final plateData = json.decode(platePrefData);
      selectedWeights = plateData.cast<num>();
    }
    notifyListeners();
  }
  
 Future<void> toggleSelection(num x) async {
    ...
    final prefs = await SharedPreferences.getInstance();
    final plateData = json.encode(selectedWeights);
    prefs.setString(PREFS_PLATES, plateData);
  }
```

in the plate widget:

```dart
class _AddPlateWeightsState extends State<AddPlateWeights> {
  @override
  void initState() {
    super.initState();

    WidgetsBinding.instance.addPostFrameCallback((_) {
      Provider.of<PlateWeights>(context, listen: false).readPlates();
    });
  }
  ...
  }

```